test-suite/testkit/tests/interfaces/main.rs

```diff
@@ -389,3 +395,156 @@ fn test_any_call_panic_recursion_limit() {
         ))
     );
 }
+
+fn execute_custom_call<F>(f: F) -> (TestKit, Result<(), ExecutionError>)
+where
+    F: Fn(ExecutionContext<'_>) -> Result<(), ExecutionError> + Clone + Send + 'static,
+{
+    let custom_call_factory = CustomCallServiceFactory::new(f);
+    let custom_call_artifact = custom_call_factory.artifact_id();
+
+    let custom_call_instance = InstanceInitParams::new(
+        CustomCallServiceFactory::INSTANCE_ID,
+        CustomCallServiceFactory::INSTANCE_NAME,
+        custom_call_artifact.clone(),
+        vec![],
+    );
+
+    let mut testkit = TestKitBuilder::validator()
+        .with_logger()
+        .with_default_rust_service(WalletService)
+        .with_artifact(custom_call_artifact)
+        .with_rust_service(custom_call_factory)
+        .with_instance(custom_call_instance)
+        .create();
+
+    let keypair = crypto::gen_keypair();
+    let res = execute_transaction(
+        &mut testkit,
+        keypair.custom_call(CustomCallServiceFactory::INSTANCE_ID, vec![]),
+    );
+    (testkit, res)
+}
+
+fn assert_access_blockchain_data(res: Result<(), ExecutionError>) {
+    assert_eq!(
+        res.unwrap_err(),
+        ErrorMatch::any_unexpected().with_description_containing(
+            "An attempt to access blockchain data after execution error"
+        )
+    );
+}
+
+#[test]
+fn custom_call_ok() {
+    execute_custom_call(|context| {
+        context.service_data();
+        Ok(())
+    })
+    .1
```

Comment:
Nit: I'd prefer destructuring assignment instead of accessing fields in a tuple.